### PR TITLE
GH-749 Tellraw multiple messages

### PR DESCRIPTION
**Pull request by CitralFlo**

### CODE REVIEW

**Comment by CitralFlo** [comment]
Due to LiteCommands not supporting sth like `@Separator` this feature has been added in few commands instead of one

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawCommand.java [hunk 1]
```diff
@@ -56,4 +61,85 @@ void tellRawAll(@Context Viewer sender, @Arg NoticeTextType type, @Join String m
...
+        this.noticeService.create()
+            .notice(translation -> translation.chat().tellrawSaved())
+            .placeholder("{MESSAGE}", message)
+            .placeholder("{TYPE}", type.name())
+            .viewer(sender)
+            .send();
+    }
+
+    @Execute(name = "-send-all")
+    @DescriptionDocs(description = "")
```
Comment (imDMK) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawCommand.java:80]: no description

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawCommand.java [hunk 2]
```diff
@@ -56,4 +61,85 @@ void tellRawAll(@Context Viewer sender, @Arg NoticeTextType type, @Join String m
...
+        this.multipleNotices.clear();
+
+        this.noticeService.create()
+            .notice(translation -> translation.chat().tellrawMultipleSent())
+            .viewer(sender)
+            .send();
+    }
+
+    @Execute(name = "-send")
+    @DescriptionDocs(description = "")
```
Comment (imDMK) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawCommand.java:107]: no description

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawCommand.java [hunk 3]
```diff
@@ -56,4 +61,85 @@ void tellRawAll(@Context Viewer sender, @Arg NoticeTextType type, @Join String m
...
+                .notice(translation -> translation.chat().tellrawNoSaved())
+                .placeholder("{PLAYER}", target.getName())
+                .viewer(sender)
+                .send();
+            return;
+        }
+
+        for (TellRawNotice notice : this.multipleNotices) {
+            this.noticeService.create()
+                .notice(notice.getNoticeTextType(), notice.getNoticeText())
```
Comment (imDMK) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawCommand.java:121]:
```suggestion
        for (TellRawNotice rawNotice : this.multipleNotices) {
            this.noticeService.create()
                .notice(rawNotice.getNoticeTextType(), rawNotice.getNoticeText())
```

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawNotice.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.eternalcode.core.feature.essentials.tellraw;
+
+import com.eternalcode.core.notice.NoticeTextType;
+import lombok.Getter;
+
+@Getter
+public class TellRawNotice {
+
+    private NoticeTextType noticeTextType;
```
Comment (imDMK) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawNotice.java:9]:
```suggestion
    private NoticeTextType noticeType;
```

**Review by igoyek — APPROVED** [review]

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawCommand.java [hunk 5]
```diff
@@ -77,25 +79,26 @@ void tellRawMultiple(@Context Viewer sender, @Arg NoticeTextType type, @Join Str
     }
 
     @Execute(name = "-send-all")
-    @DescriptionDocs(description = "")
+    @DescriptionDocs(description = "Send all saved messages to all players")
     void tellRawSend(@Context Viewer sender) {
+        UUID uuid = sender.getUniqueId();
 
-        if (this.multipleNotices.isEmpty()) {
+        if (this.tellRawService.getNotices(uuid) == null || this.tellRawService.getNotices(uuid).isEmpty()) {
```
Comment (imDMK) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawCommand.java:86]: create a separate method in TellRawService, maybe named ```hasNotices``` and return boolean

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawService.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+    void addNotice(UUID uuid, TellRawNotice notice) {
+        Set<TellRawNotice> notices = this.multipleNotices.get(uuid);
+
+        if (notices == null) {
+            notices = new HashSet<>();
+        }
+
+        notices.add(notice);
+        this.multipleNotices.put(uuid, notices);
```
Comment (imDMK) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawService.java:27]:
```suggestion
        Set<TellRawNotice> notices = this.multipleNotices.computeIfAbsent(uuid, k -> new HashSet<>());
        
        notices.add(notice);
```

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawNotice.java [hunk 7]
```diff
@@ -1,16 +1,6 @@
 package com.eternalcode.core.feature.essentials.tellraw;
 
 import com.eternalcode.core.notice.NoticeTextType;
-import lombok.Getter;
 
-@Getter
-public class TellRawNotice {
+record TellRawNotice(NoticeTextType noticeTextType, String message) {}
```
Comment (imDMK) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawNotice.java:5]:
```suggestion
record TellRawNotice(NoticeTextType noticeType, String message) {}
```

**Review by imDMK — CHANGES_REQUESTED** [review]

**Review by vLuckyyy — COMMENTED** [review]
Hmm? 

maybe:
/tellraw -single <type>  <text>
/tellraw -multiple - star's notice queue
/tellraw -multiple-send

imo the above format is better readable. I'm opening discussions, have your say

**Comment by CitralFlo** [comment]
> Hmm?
> 
> maybe: /tellraw **-single** `<type>` `<text>` /tellraw **-multiple** - star's notice queue /tellraw **-multiple-send** - sends notices from queue, but you can set up hover message with click event to make this action quicker
> 
> imo the above format is better readable. I'm opening discussions, have your say

I was thinking about NOT changing current command usage, so letting it be  like `/tellraw * chat <message>`
The other suggestions look fine, so i can stick with them.

**Comment by CitralFlo** [comment]
New command structure: 
image

image

**Review by Jakubk15 — APPROVED** [review]

**Review by imDMK — APPROVED** [review]

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawMultipleCommand.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+        this.noticeService.create()
+            .notice(translation -> translation.chat().tellrawMultipleSent())
+            .viewer(sender)
+            .send();
+    }
+
+    @Execute(name = "-send")
+    @DescriptionDocs(description = "Send all saved messages to the player", arguments = "<player>")
+    void tellRawSend(@Context Viewer sender, @Arg Player target) {
```
Comment (Rollczi) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/essentials/tellraw/TellRawMultipleCommand.java:75]: Nie rozumiem, czemu są chyba dwie identyczne metody? Tylko jedna to send a drugie to all? Nie lepiej dać po prostu send?

**Review by Rollczi — CHANGES_REQUESTED** [review]

**Comment by Rollczi** [comment]
Ja bym trochę przerobił parę komend.
Na coś w stylu:

/tellraw quote add <type> <content>
/tellraw quote clear
/tellraw quote send-all
/tellraw quote send <count>

Ogólnie myślniki używa się jako flagi ale subkomendy to średnio.

Np.
/chat off
/chat on
/chat slowmode

**Comment by CitralFlo** [comment]
> quote

Firstly I don't like your proposition.
My proposition:

image

Your proposition (vLuckyyy corrected that You meant "queue"):
image

Problem with yours will be when there are more players on the server and You cannot see "queue" part.

image

**Review by Rollczi — APPROVED** [review]

**Pull request merged**